URBNSwiftyConvenience/Classes/URBNBorders.swift

```diff
@@ -0,0 +1,290 @@
+                borderView.trailingAnchor.constraint(equalTo: trailingAnchor).isActive = true
+                borderView.topAnchor.constraint(equalTo: topAnchor).isActive = true
+                borderView.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
+            }
+            else {
+                let views = ["borderView" : borderView]
+                activateVFL(format: "V:|[borderView]|", options: .alignAllCenterX, metrics: nil, views: views)
+                activateVFL(format: "H:|[borderView]|", options: .alignAllCenterY, metrics: nil, views: views)
+            }
+
+            objc_setAssociatedObject(self, &kURBNBorderViewKey, borderView, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+            
+            return borderView
+        }
+        
+        return nil
+    }
+
+    // MARK: Getters
+    private func leftBorder() -> URBNBorder? {
+        return self.urbn_borderView()?.urbn_LeftBorder()
+    }
+    
+    private func rightBorder() -> URBNBorder? {
+        return self.urbn_borderView()?.urbn_RightBorder()
+    }
+    
+    private func topBorder() -> URBNBorder? {
+        return self.urbn_borderView()?.urbn_TopBorder()
+    }
+    
+    private func bottomBorder() -> URBNBorder? {
+        return self.urbn_borderView()?.urbn_BottomBorder()
+    }
+}
+
+public class URBNBorder: NSObject {
+    public var color = UIColor()
+    public var width = CGFloat()
+    public let insets = UIEdgeInsets()
+}
+
+fileprivate final class URBNBorderView: UIView {
+    fileprivate var bv_leftBorder: URBNBorder?
+    fileprivate var bv_rightBorder: URBNBorder?
+    fileprivate var bv_topBorder: URBNBorder?
+    fileprivate var bv_bottomBorder: URBNBorder?
+    
+    private final let width = "width"
+    private final let color = "color"
+    private final let insets = "insets"
+    
+    convenience init() {
+        self.init()
+    }
+    
+    override convenience init(frame: CGRect) {
+        self.init(frame: frame)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    // MARK: KVO
+    private func configuredBorder() -> URBNBorder {
+        let border = URBNBorder()
+        border.addObserver(self, forKeyPath: width, options: .new, context: nil)
+        border.addObserver(self, forKeyPath: color, options: .new, context: nil)
+        border.addObserver(self, forKeyPath: insets, options: .new, context: nil)
+        
+        return border
+    }
+    
+    private func unregisterKVO(forBorder border: URBNBorder) {
+        border.removeObserver(self, forKeyPath: width)
+        border.removeObserver(self, forKeyPath: color)
+        border.removeObserver(self, forKeyPath: insets)
+    }
+    
+    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
+        setNeedsDisplay()
```

Comment:
Ugh. So, the original code is a mess. This is wrong too. Let me see if I can explain:

You’re overriding `observeValue(forKeyPath…)`. UIView’s inheritance hierarchy is:
1. NSObject
2. UIResponder
3. UIView

So if the UIView, UIResponder, or NSObject add observers it’ll get rerouted through here. Unfortunately, you’re not calling the superclass `observeValue(forKeyPath` so those messages won't be handled. The general logic for `observeValue(forKeyPath…` is to check the context. If it’s of our know type, handle it, otherwise call `super.observeValue(forKeyPath…`.

But…

You’re overriding a method in an extension on the same class. [This is incorrect and Apple says not to do it](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/CustomizingExistingClasses/CustomizingExistingClasses.html#//apple_ref/doc/uid/TP40011210-CH6-SW4). If Apple puts any custom logic in `observeValue(forKeyPath…` in UIView, then the result is undefined. The Objective-C runtime may call this method and completely ignore Apple. Or it make call Apple’s an completely ignore this.

If this we’re straight Swift, the compiler wouldn't even allow overriding a function that’s already been overridden in the same class. Because this is UIView, it’s implicitly `@objc` which makes this allowed.

Does this make sense? I think the ultimate solution is to get rid of the KVO. I’m just not sure we can use swift value observation as a replacement, though:

```swift
var urbn_topBorder: URBNBorder? {
    didSet {
        setNeedsDisplay()
    }
}
```